File: src/main/java/com/tkr/desk/CorporateActionAdjuster.java
package com.tkr.desk;

import com.tkr.types.WireTypes.*;
import com.tkr.types.WireTypes.Status;

import java.util.ArrayList;
import java.util.List;

/**
 * Corporate action adjuster for splits, dividends, mergers, spin-offs.
 * Ported from legacy-cpp corporate_action_adjuster.cc.
 */
public final class CorporateActionAdjuster {

    public enum CorporateActionKind {
        STOCK_SPLIT, REVERSE_SPLIT, CASH_DIVIDEND, STOCK_DIVIDEND, MERGER, SPIN_OFF
    }

    public static final class CorporateAction {
        public int actionId;
        public CorporateActionKind kind;
        public int symbolId;
        public int newSymbolId;
        public long ratioNumerator;
        public long ratioDenominator;
        public long cashAmountCents;
    }

    public static final class AdjustedPosition {
        public int accountId;
        public int symbolId;
        public long oldQtyMilli;
        public long newQtyMilli;
        public long cashAdjustmentCents;
        public int actionId;
    }

    public static final class CorporateActionAdjusterConfig {
        public int deskId;

        public CorporateActionAdjusterConfig(int deskId) {
            this.deskId = deskId;
        }
    }

    public static final class CorporateActionAdjusterResult {
        public Status status = Status.OK;
        public List<AdjustedPosition> adjustments = new ArrayList<>();
        public long totalCashCents;
        public int actionsApplied;
    }

    /** Legacy wrapper. */
    public static final class AdjustResult {
        public Status status = Status.OK;
        public List<AdjustedPosition> adjustments = new ArrayList<>();
    }

    private final CorporateActionAdjusterConfig config;
    private final List<CorporateAction> pendingActions = new ArrayList<>();

    public CorporateActionAdjuster() {
        this(new CorporateActionAdjusterConfig(0));
    }

    public CorporateActionAdjuster(CorporateActionAdjusterConfig config) {
        this.config = config != null ? config : new CorporateActionAdjusterConfig(0);
    }

    public void registerAction(CorporateAction action) {
        if (action != null) pendingActions.add(action);
    }

    public void clearActions() {
        pendingActions.clear();
    }

    public static long adjustQtyForSplit(long qtyMilli, long numerator, long denominator) {
        if (denominator == 0) return qtyMilli;
        return qtyMilli * numerator / denominator;
    }

    public static long computeCashDividend(long qtyMilli, long cashPerShareCents) {
        return qtyMilli * cashPerShareCents / 1000L;
    }

    public AdjustedPosition applyStockSplit(CorporateAction action, PositionLedger.PositionEntry pos) {
        AdjustedPosition adj = new AdjustedPosition();
        adj.accountId = pos.accountId;
        adj.symbolId = pos.symbolId;
        adj.oldQtyMilli = pos.qtyMilli;
        adj.newQtyMilli = adjustQtyForSplit(pos.qtyMilli, action.ratioNumerator, action.ratioDenominator);
        adj.cashAdjustmentCents = 0;
        adj.actionId = action.actionId;
        return adj;
    }

    public AdjustedPosition applyReverseSplit(CorporateAction action, PositionLedger.PositionEntry pos) {
        return applyStockSplit(action, pos);
    }

    public AdjustedPosition applyCashDividend(CorporateAction action, PositionLedger.PositionEntry pos) {
        AdjustedPosition adj = new AdjustedPosition();
        adj.accountId = pos.accountId;
        adj.symbolId = pos.symbolId;
        adj.oldQtyMilli = pos.qtyMilli;
        adj.newQtyMilli = pos.qtyMilli;
        adj.cashAdjustmentCents = computeCashDividend(pos.qtyMilli, action.cashAmountCents);
        adj.actionId = action.actionId;
        return adj;
    }

    public AdjustedPosition applyStockDividend(CorporateAction action, PositionLedger.PositionEntry pos) {
        AdjustedPosition adj = applyStockSplit(action, pos);
        adj.cashAdjustmentCents = 0;
        return adj;
    }

    public AdjustedPosition applyMerger(CorporateAction action, PositionLedger.PositionEntry pos) {
        AdjustedPosition adj = new AdjustedPosition();
        adj.accountId = pos.accountId;
        adj.symbolId = action.newSymbolId;
        adj.oldQtyMilli = pos.qtyMilli;
        adj.newQtyMilli = adjustQtyForSplit(pos.qtyMilli, action.ratioNumerator, action.ratioDenominator);
        adj.cashAdjustmentCents = action.cashAmountCents;
        adj.actionId = action.actionId;
        return adj;
    }

    public AdjustedPosition applySpinOff(CorporateAction action, PositionLedger.PositionEntry pos) {
        AdjustedPosition adj = new AdjustedPosition();
        adj.accountId = pos.accountId;
        adj.symbolId = action.newSymbolId;
        adj.oldQtyMilli = pos.qtyMilli;
        adj.newQtyMilli = adjustQtyForSplit(pos.qtyMilli, action.ratioNumerator, action.ratioDenominator);
        adj.cashAdjustmentCents = 0;
        adj.actionId = action.actionId;
        return adj;
    }

    public CorporateActionAdjusterResult applyActions(List<PositionLedger.PositionEntry> positions) {
        CorporateActionAdjusterResult result = new CorporateActionAdjusterResult();
        result.status = Status.OK;
        for (PositionLedger.PositionEntry pos : positions) {
            for (CorporateAction action : pendingActions) {
                if (action.symbolId != pos.symbolId) continue;
                AdjustedPosition adj = switch (action.kind) {
                    case STOCK_SPLIT -> applyStockSplit(action, pos);
                    case REVERSE_SPLIT -> applyReverseSplit(action, pos);
                    case CASH_DIVIDEND -> applyCashDividend(action, pos);
                    case STOCK_DIVIDEND -> applyStockDividend(action, pos);
                    case MERGER -> applyMerger(action, pos);
                    case SPIN_OFF -> applySpinOff(action, pos);
                };
                result.adjustments.add(adj);
                result.totalCashCents += adj.cashAdjustmentCents;
                result.actionsApplied++;
            }
        }
        return result;
    }

    public CorporateActionAdjusterResult applyToBatch(BatchWireFrame frame) {
        List<PositionLedger.PositionEntry> positions = new ArrayList<>();
        for (WireBatchRecord rec : frame.records) {
            PositionLedger.PositionEntry pos = new PositionLedger.PositionEntry();
            pos.accountId = rec.accountId;
            pos.symbolId = rec.symbolId;
            pos.qtyMilli = rec.qtyMilli;
            positions.add(pos);
        }
        return applyActions(positions);
    }

    /** Legacy API. */
    public AdjustResult adjustBatch(BatchWireFrame frame) {
        CorporateActionAdjusterResult engine = applyToBatch(frame);
        AdjustResult legacy = new AdjustResult();
        legacy.status = engine.status;
        legacy.adjustments = engine.adjustments;
        return legacy;
    }

    public List<CorporateAction> pendingActions() {
        return new ArrayList<>(pendingActions);
    }
}

File: legacy-cpp/src/desk/corporate_action_adjuster.cc
#include "tkr/desk/corporate_action_adjuster.h"

namespace tkr {
namespace desk {

CorporateActionAdjuster::CorporateActionAdjuster(
    CorporateActionAdjusterConfig config)
    : config_(config) {}

void CorporateActionAdjuster::RegisterAction(const CorporateAction& action) {
  pending_actions_.push_back(action);
}

void CorporateActionAdjuster::ClearActions() { pending_actions_.clear(); }

std::int64_t CorporateActionAdjuster::AdjustQtyForSplit(
    std::int64_t qty_milli, std::int64_t numerator,
    std::int64_t denominator) {
  if (denominator == 0) {
    return qty_milli;
  }
  return (qty_milli * numerator) / denominator;
}

std::int64_t CorporateActionAdjuster::ComputeCashDividend(
    std::int64_t qty_milli, std::int64_t cash_per_share_cents) {
  return (qty_milli * cash_per_share_cents) / 1000;
}

AdjustedPosition CorporateActionAdjuster::ApplyStockSplit(
    const CorporateAction& action, const PositionEntry& pos) const {
  AdjustedPosition adj{};
  adj.account_id = pos.account_id;
  adj.symbol_id = pos.symbol_id;
  adj.old_qty_milli = pos.qty_milli;
  adj.new_qty_milli =
      AdjustQtyForSplit(pos.qty_milli, action.ratio_numerator,
                        action.ratio_denominator);
  adj.cash_adjustment_cents = 0;
  adj.action_id = action.action_id;
  return adj;
}

AdjustedPosition CorporateActionAdjuster::ApplyReverseSplit(
    const CorporateAction& action, const PositionEntry& pos) const {
  AdjustedPosition adj = ApplyStockSplit(action, pos);
  adj.action_id = action.action_id;
  return adj;
}

AdjustedPosition CorporateActionAdjuster::ApplyCashDividend(
    const CorporateAction& action, const PositionEntry& pos) const {
  AdjustedPosition adj{};
  adj.account_id = pos.account_id;
  adj.symbol_id = pos.symbol_id;
  adj.old_qty_milli = pos.qty_milli;
  adj.new_qty_milli = pos.qty_milli;
  adj.cash_adjustment_cents =
      ComputeCashDividend(pos.qty_milli, action.cash_amount_cents);
  adj.action_id = action.action_id;
  return adj;
}

AdjustedPosition CorporateActionAdjuster::ApplyStockDividend(
    const CorporateAction& action, const PositionEntry& pos) const {
  AdjustedPosition adj = ApplyStockSplit(action, pos);
  adj.cash_adjustment_cents = 0;
  return adj;
}

AdjustedPosition CorporateActionAdjuster::ApplyMerger(
    const CorporateAction& action, const PositionEntry& pos) const {
  AdjustedPosition adj{};
  adj.account_id = pos.account_id;
  adj.symbol_id = action.new_symbol_id;
  adj.old_qty_milli = pos.qty_milli;
  adj.new_qty_milli =
      AdjustQtyForSplit(pos.qty_milli, action.ratio_numerator,
                        action.ratio_denominator);
  adj.cash_adjustment_cents = action.cash_amount_cents;
  adj.action_id = action.action_id;
  return adj;
}

AdjustedPosition CorporateActionAdjuster::ApplySpinOff(
    const CorporateAction& action, const PositionEntry& pos) const {
  AdjustedPosition adj{};
  adj.account_id = pos.account_id;
  adj.symbol_id = action.new_symbol_id;
  adj.old_qty_milli = pos.qty_milli;
  adj.new_qty_milli =
      AdjustQtyForSplit(pos.qty_milli, action.ratio_numerator,
                        action.ratio_denominator);
  adj.cash_adjustment_cents = 0;
  adj.action_id = action.action_id;
  return adj;
}

CorporateActionAdjusterResult CorporateActionAdjuster::ApplyActions(
    const std::vector<PositionEntry>& positions) {
  CorporateActionAdjusterResult result{};
  result.status = Status::kOk;

  for (const PositionEntry& pos : positions) {
    for (const CorporateAction& action : pending_actions_) {
      if (action.symbol_id != pos.symbol_id) {
        continue;
      }

      AdjustedPosition adj{};
      switch (action.kind) {
        case CorporateActionKind::kStockSplit:
          adj = ApplyStockSplit(action, pos);
          break;
        case CorporateActionKind::kReverseSplit:
          adj = ApplyReverseSplit(action, pos);
          break;
        case CorporateActionKind::kCashDividend:
          adj = ApplyCashDividend(action, pos);
          break;
        case CorporateActionKind::kStockDividend:
          adj = ApplyStockDividend(action, pos);
          break;
        case CorporateActionKind::kMerger:
          adj = ApplyMerger(action, pos);
          break;
        case CorporateActionKind::kSpinOff:
          adj = ApplySpinOff(action, pos);
          break;
        default:
          continue;
      }

      result.adjustments.push_back(adj);
      result.total_cash_cents += adj.cash_adjustment_cents;
      ++result.actions_applied;
    }
  }

  return result;
}

CorporateActionAdjusterResult CorporateActionAdjuster::ApplyToBatch(
    const BatchWireFrame& frame) {
  std::vector<PositionEntry> positions;
  positions.reserve(frame.records.size());

  for (const WireBatchRecord& rec : frame.records) {
    PositionEntry pos{};
    pos.account_id = rec.account_id;
    pos.symbol_id = rec.symbol_id;
    pos.qty_milli = static_cast<std::int64_t>(rec.qty_milli);
    positions.push_back(pos);
  }

  return ApplyActions(positions);
}

}  // namespace desk
}  // namespace tkr
